RemoveNAindv.GCL.R

```diff
@@ -1,47 +1,60 @@
-#
-#  Argument(s):  
-#  sillyvec <- character vector of SILLYs
-#
-#  Output:
-#  1) this function will modify silly.gcls in your environment to remove NA fish
-#  2) output is a list of FK_FISH_ID of fish removed for each SILLY
-#
-#  Written by Kyle Shedd 10/16/15  
-#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-  na.individuals.to.remove <- sapply(sillyvec, function(silly) {
-    my.gcl <- get(paste(silly, ".gcl", sep = ""))
-    my.gcl$attributes$FK_FISH_ID[apply(my.gcl$scores, 1, function(ind) {
-      sum(is.na(ind[, 1])) > 0 
-    })]
-  }, simplify = FALSE, USE.NAMES = TRUE)
   
-  sapply(sillyvec, function(silly) {
-    if(length(na.individuals.to.remove[[silly]]) > 0){
-      my.gcl <- get(paste(silly, ".gcl", sep = ""))
-      counts <- my.gcl$counts
-      scores <- my.gcl$scores
-      n <- my.gcl$n
-      attributes <- my.gcl$attributes
-      
-      ind.to.keep <- !my.gcl$attributes$FK_FISH_ID %in% na.individuals.to.remove[[silly]]
+  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  #   This function removes any individuals from a SILLY.gcl object that were not genotyped
+  #   for all markers in markersuite.
+  #
+  # Inputs~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  #   
+  #   sillyvec - a character vector of silly codes (e.g. sillyvec <- c("KQUART06","KQUART08","KQUART09")) 
+  # 
+  # Outputs~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  #    1) this function will modify silly.gcls in your environment to remove NA fish
+  #    2) output is tibble of the fish IDs removed by silly
+  #
+  # Example~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  #   load("V:/Analysis/2_Central/Chinook/Cook Inlet/2019/2019_UCI_Chinook_baseline_hap_data/2019_UCI_Chinook_baseline_hap_data.RData")
+  # 
+  #   removedInd <- RemoveNAindv.GCL(sillyvec = sillyvec157)
+  #
+  # Note~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  # This function requires a locus control object to subset out the scores columns.
+  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  
+  if(!exists("LocusControl")){
+    
+    stop("'LocusControl' not yet built.")
+    
+  }
+  
+  if(!require("pacman")) install.packages("pacman"); library(pacman); pacman::p_load(tidyverse) #Install packages, if not in library and then load them.
+  
+  scores_cols <- with(LocusControl, c(locusnames, paste0(LocusControl$locusnames, ".1"))) %>% 
+    sort()
+  
+  na.individuals.removed <- lapply(sillyvec, function(silly) {
+    
+    my.gcl <- get(paste(silly, ".gcl", sep = ""))
+    
+    drop <- my.gcl$FK_FISH_ID[apply(my.gcl[ , scores_cols], 1, function(ind) {
       
-      counts <- counts[ind.to.keep, , ,drop=FALSE]
-      scores <- scores[ind.to.keep, , ,drop=FALSE]
-      n <- nrow(scores)
-      attributes <- attributes[ind.to.keep, ,drop=FALSE]
+      sum(is.na(ind)) > 0 
       
-      assign(paste(silly, ".gcl", sep = ""), list(counts = counts, scores = scores, n = n, attributes = attributes), pos = 1)
-    }} )
+    })]
```

Comment:
So this actually ends up dropping any individual that is missing data for any reason (no calls and never genotyped), because we convert all missing genotypes to NA in LOKI2R. In the old format of .gcl objects, the scores array had NAs for never genotyped and 0s for no calls. This function would then look for individuals with NAs (i.e. never genotyped for a given locus). Maybe we should just retire this function and fold it in with LOKI2R since we already calculate missing_loci? We can use that to cull individuals that were not genotyped for all loci.

Reply:
I figured out what the problem was here, we were looking at allele_1 and allele_2. All haploid/mitochondrial markers are default NA for allele_2, so if your markersuite contains any haploid markers, all individuals are removed. Just look at allele_1.